cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -235,7 +237,23 @@ public ArtifactDetail getArtifact(Id.Artifact artifactId) throws IOException, Ar
   public Map.Entry<ArtifactDescriptor, PluginClass> findPlugin(Id.Artifact artifactId, String pluginType,
                                                                String pluginName, PluginSelector selector)
     throws IOException, PluginNotExistsException {
-    return selector.select(artifactStore.getPluginClasses(artifactId, pluginType, pluginName));
+    SortedMap<ArtifactDescriptor, PluginClass> pluginClasses = artifactStore.getPluginClasses(
+      artifactId, pluginType, pluginName);
+    SortedMap<ArtifactId, PluginClass> artifactIds = Maps.newTreeMap();
+    for (Map.Entry<ArtifactDescriptor, PluginClass> pluginClassEntry : pluginClasses.entrySet()) {
+      artifactIds.put(pluginClassEntry.getKey().getArtifactId(), pluginClassEntry.getValue());
+    }
+    Map.Entry<ArtifactId, PluginClass> chosenArtifact = selector.select(artifactIds);
+    if (chosenArtifact == null) {
+      throw new PluginNotExistsException(artifactId, pluginType, pluginName);
+    }
+
+    for (Map.Entry<ArtifactDescriptor, PluginClass> pluginClassEntry : pluginClasses.entrySet()) {
```

Comment:
we can't just call get on the map? ArtifactDescriptor doesn't use location in its equality and comparison implementations right?